### ISSUE TITLE
2.0 - Cannot build on OSX

### ISSUE THREAD

**Issue by jeremyabel** [KhronosGroup/COLLADA2GLTF#91]
See: https://github.com/lasalvavida/ahoy/issues/1
@lasalvavida

**Comment by jeremyabel** [comment]
pull request fix: https://github.com/lasalvavida/ahoy/pull/2

**Comment by jeremyabel** [comment]
With my patch for Ahoy seeming to work, the build process is currently erroring out after libXML (OpenCOLLADA dependency) can't find malloc.h. Malloc.h on OSX needs to be handled specially, but I'm not sure of the details. This doesn't happen on the master branch though, OpenCOLLADA compiles fine there. I've pasted the command line output from `make` below:

```
Jeremys-iMac:build jeremyabel$ make
Scanning dependencies of target ahoy
[  0%] Building CXX object dependencies/ahoy/CMakeFiles/ahoy.dir/src/Parser.cpp.o
clang: warning: -lstdc++fs: 'linker' input unused
[  1%] Building CXX object dependencies/ahoy/CMakeFiles/ahoy.dir/src/TypedOption.cpp.o
clang: warning: -lstdc++fs: 'linker' input unused
/Users/jeremyabel/Documents/Workspace/COLLADA2GLTF/dependencies/ahoy/src/TypedOption.cpp:121:70: warning: adding 'int' to a string does not append to the string [-Wstring-plus-int]
                return Option::help() + " [vec<int>" + (countValue < 0 ? "" : ", " + countValue) + "]" + (hasDefault ? (" [default: " + defaultString + "]") : "");
                                                                              ~~~~~^~~~~~~~~~~~
/Users/jeremyabel/Documents/Workspace/COLLADA2GLTF/dependencies/ahoy/src/TypedOption.cpp:121:70: note: use array indexing to silence this warning
                return Option::help() + " [vec<int>" + (countValue < 0 ? "" : ", " + countValue) + "]" + (hasDefault ? (" [default: " + defaultString + "]") : "");
                                                                                   ^
                                                                              &    [           ]
1 warning generated.
[  1%] Linking CXX static library libahoy.a
[  1%] Built target ahoy
Scanning dependencies of target LibXML
[  1%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/c14n.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  1%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/catalog.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  2%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/chvalid.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  2%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/debugXML.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  2%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/dict.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  2%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/DOCBparser.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  3%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/encoding.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  3%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/entities.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  3%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/error.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  3%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/legacy.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  4%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/list.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  4%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/parser.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  4%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/parserInternals.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  5%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/pattern.c.o
clang: warning: -lstdc++fs: 'linker' input unused
/Users/jeremyabel/Documents/Workspace/COLLADA2GLTF/dependencies/OpenCOLLADA/OpenCOLLADA/Externals/LibXML/pattern.c:308:43: warning: for loop has empty body [-Wempty-body]
        for (i = 0;namespaces[2 * i] != NULL;i++);
                                                 ^
/Users/jeremyabel/Documents/Workspace/COLLADA2GLTF/dependencies/OpenCOLLADA/OpenCOLLADA/Externals/LibXML/pattern.c:308:43: note: put the semicolon on a separate line to silence this warning
1 warning generated.
[  5%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/globals.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  5%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/hash.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  5%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/HTMLparser.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  6%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/HTMLtree.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  6%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/SAX2.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  6%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/SAX.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  6%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/threads.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  7%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/relaxng.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  7%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/uri.c.o
clang: warning: -lstdc++fs: 'linker' input unused
/Users/jeremyabel/Documents/Workspace/COLLADA2GLTF/dependencies/OpenCOLLADA/OpenCOLLADA/Externals/LibXML/uri.c:433:23: warning: passing 'char *' to parameter of type 'const xmlChar *' (aka 'const unsigned char *') converts between pointers to integer types with
      different sign [-Wpointer-sign]
                (xmlStrEqual(uri->scheme, BAD_CAST "file"))) {
                             ^~~~~~~~~~~
/Users/jeremyabel/Documents/Workspace/COLLADA2GLTF/dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/../../../OpenCOLLADA/Externals/LibXML/include/libxml/xmlstring.h:78:58: note: passing argument to parameter 'str1' here
                xmlStrEqual              (const xmlChar *str1,
                                                         ^
/Users/jeremyabel/Documents/Workspace/COLLADA2GLTF/dependencies/OpenCOLLADA/OpenCOLLADA/Externals/LibXML/uri.c:797:33: warning: while loop has empty body [-Wempty-body]
        while ((*tmp++ = *segp++) != 0);
                                       ^
/Users/jeremyabel/Documents/Workspace/COLLADA2GLTF/dependencies/OpenCOLLADA/OpenCOLLADA/Externals/LibXML/uri.c:797:33: note: put the semicolon on a separate line to silence this warning
2 warnings generated.
[  7%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/valid.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  7%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/xinclude.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  8%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/xlink.c.o
clang: warning: -lstdc++fs: 'linker' input unused
[  8%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/xmlIO.c.o
clang: warning: -lstdc++fs: 'linker' input unused
/Users/jeremyabel/Documents/Workspace/COLLADA2GLTF/dependencies/OpenCOLLADA/OpenCOLLADA/Externals/LibXML/xmlIO.c:768:11: warning: implicit declaration of function 'read' is invalid in C99 [-Wimplicit-function-declaration]
    ret = read((int) (long) context, &buffer[0], len);
          ^
/Users/jeremyabel/Documents/Workspace/COLLADA2GLTF/dependencies/OpenCOLLADA/OpenCOLLADA/Externals/LibXML/xmlIO.c:807:11: warning: implicit declaration of function 'close' is invalid in C99 [-Wimplicit-function-declaration]
    ret = close((int) (long) context);
          ^
/Users/jeremyabel/Documents/Workspace/COLLADA2GLTF/dependencies/OpenCOLLADA/OpenCOLLADA/Externals/LibXML/xmlIO.c:3535:13: warning: implicit declaration of function 'getcwd' is invalid in C99 [-Wimplicit-function-declaration]
        if (getcwd(dir, 1024) != NULL) {
            ^
/Users/jeremyabel/Documents/Workspace/COLLADA2GLTF/dependencies/OpenCOLLADA/OpenCOLLADA/Externals/LibXML/xmlIO.c:3535:31: warning: comparison between pointer and integer ('int' and 'void *')
        if (getcwd(dir, 1024) != NULL) {
            ~~~~~~~~~~~~~~~~~ ^  ~~~~
4 warnings generated.
[  8%] Building C object dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/xmlmemory.c.o
clang: warning: -lstdc++fs: 'linker' input unused
/Users/jeremyabel/Documents/Workspace/COLLADA2GLTF/dependencies/OpenCOLLADA/OpenCOLLADA/Externals/LibXML/xmlmemory.c:24:10: fatal error: 'malloc.h' file not found
#include <malloc.h>
         ^
1 error generated.
make[2]: *** [dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/__/__/__/__/OpenCOLLADA/Externals/LibXML/xmlmemory.c.o] Error 1
make[1]: *** [dependencies/OpenCOLLADA/modules/COLLADASaxFrameworkLoader/GeneratedSaxParser/LibXML/CMakeFiles/LibXML.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by jeremyabel** [comment]
Manually commenting out the line `#define HAVE_MALLOC_H` in config.h in OpenCOLLADA's LibXML seems to do the trick for now. I'm sure there's a way for CMake to do this but I'm not too knowledgeable about CMake in general. There's a mac-specific config file in `LibXML/macos/scr/config-mac.h` as well. 

Currently I've hit a wall at the fact that `GLTFOptions.h` uses `std::experimental::filesystem` which seems to be part of a newer version of gcc than what comes installed on the latest version of OSX.

**Comment by lasalvavida** [comment]
See also #72, #14

**Comment by lasalvavida** [comment]
> Currently I've hit a wall at the fact file I/O was updated to use std::experimental::filesystem::path which seems to be part of a newer version of gcc than what comes installed on the latest version of OSX.

Yes, this is also a problem for building with emscripten #58. `std:::experimental::filesystem` is API compatible with boost::filesystem, but I'm not crazy about trying to integrate boost and either complicating the build system terribly or requiring devs to have boost already installed.

I'm thinking the solution is probably going to be to write a shim filesystem library. I've looked around a bit and some do exist but I haven't found one whose quality I am satisfied with. If you see one that you think looks promising feel free to link it here.

**Comment by jeremyabel** [comment]
I'm definitely in agreement that integrating or requiring boost would get sticky real quick. Here's one quick solution: 

```
$ brew install -v gcc --HEAD
$ mkdir build
$ cd build
$ export CC=/usr/local/bin/gcc-8
$ export CXX=/usr/local/bin/g++-8
$ cmake ..
$ make
```

Just manually install gcc and use that instead! Works great, the build completed just fine.

**Comment by radames** [comment]
@jeremyabel following your steps it has compiled perfect on my OSX(10.12.6)

**Comment by liuquanzhang** [comment]
Where is the comment in #define HAVE_MALLOC_HLibXML

**Comment by liuquanzhang** [comment]
1 error generated.
make[2]: *** [GLTF/CMakeFiles/GLTF.dir/src/GLTFAccessor.cpp.o] Error 1
make[1]: *** [GLTF/CMakeFiles/GLTF.dir/all] Error 2
make: *** [all] Error 2

**Comment by lasalvavida** [comment]
This should be fixed as of v2.1.3. Can anyone with a Mac confirm?

**Comment by lasalvavida** [comment]
There are also pre-built releases for Mac OS under v2.1.3 and moving forward.
